doc/man/cbc-n1qlback.1
.\" generated with Ronn-NG/v0.10.1
.\" http://github.com/apjanke/ronn-ng/tree/0.10.1
.TH "CBC\-N1QLBACK" "1" "July 2024" ""
.SH "NAME"
\fBcbc\-n1qlback\fR \- Stress Test for Couchbase Query (N1QL)
.SH "SYNOPSIS"
\fBcbc\-n1qlback\fR \-f QUERYFILE [\fIOPTIONS\fR]
.SH "DESCRIPTION"
\fBcbc\-n1qlback\fR creates a specified number of threads each executing a set of user defined queries\.
.P
\fBcbc\-n1qlback\fR requires that it be passed the path to a file containing the queries to execute; one per line\. The query should be in the format of the actual HTTP POST body (in JSON format) to be sent to the server\. For simple queries, only the \fBstatement\fR field needs to be set:
.IP "" 4
.nf
{"statement":"SELECT country FROM `travel\-sample`"}
{"statement":"SELECT country, COUNT(country) FROM `travel\-sample` GROUP BY country"}
.fi
.IP "" 0
.P
For more complex queries (for example, placeholders, custom options), you may refer to the N1QL REST API reference\.
.P
For example, the following line shows how to use named parameter
.IP "" 4
.nf
{"statement":"SELECT RAW meta()\.id FROM `travel\-sample` WHERE type=$type LIMIT 1", "$type":"airline"}
.fi
.IP "" 0
.P
There is also special query option, \fB"n1qlback"\fR, which is stripped before sending the payload\. Currently it allows to tell if the particular query must be prepared before execution:
.IP "" 4
.nf
{"statement":"SELECT * FROM `travel\-sample` WHERE type=$type LIMIT 10", "$type":"airline", "n1qlback": {"prepare": true}}
.fi
.IP "" 0
.P
\fBn1qlback\fR requires that any resources (data items, indexes) are already defined\.
.SH "OPTIONS"
The following options control workload generation:
.TP
\fB\-f\fR \fB\-\-queryfile\fR=\fIPATH\fR
Path to a file containing the query bodies to execute in JSON format, one query per line\. See above for the format\.
.TP
\fB\-t\fR, \fB\-\-num\-threads\fR=\fINTHREADS\fR
Set the number of threads (and thus the number of client instances) to run concurrently\. Each thread is assigned its own client object\.
.P
The following options control how \fBcbc\-n1qlback\fR connects to the cluster
.P
@@common\-options\.markdown@@
.TP
\fB\-e\fR, \fB\-\-error\-log\fR=\fIPATH\fR
Path to a file, where the command will write failed queries along with error details\. Use this option to figure out why \fBERRORS\fR metric is not zero\.
.P
 \fI\fR
.SH "ADDITIONAL OPTIONS"
The following options may be included in the connection string (via the \fB\-U\fR option) as URI\-style query params (e\.g\. \fBcouchbase://host/bucket?option1=value1&option2=value2\fR) or as individual key=value pairs passed to the \fB\-D\fR switch (e\.g\. \fB\-Doption1=value1 \-Doption2=value\fR)\. The \fB\-D\fR will internally build the connection string, and is provided as a convenience for options to be easily passed on the command\-line
.P
@@common\-additional\-options\.markdown@@
.SH "EXAMPLES"
The following will create a file with 3 queries and 5 threads alternating between them\. It also creates indexes on the \fBtravel\-sample\fR bucket
.IP "" 4
.nf
cbc n1ql 'CREATE INDEX ix_name ON `travel\-sample`(name)'
cbc n1ql 'CREATE INDEX ix_country ON `travel\-sample`(country)'
.fi
.IP "" 0
.P
Crete text file \fBqueries\.txt\fR with the following content (note that fourth query has parameter)
.IP "" 4
.nf
{"statement":"SELECT country FROM `travel\-sample` WHERE `travel\-sample`\.country = \e"United States\e""}
{"statement":"SELECT name FROM `travel\-sample` LIMIT 10"}
{"statement":"SELECT country, COUNT(country) FROM `travel\-sample` GROUP BY country"}
{"statement":"SELECT RAW meta()\.id FROM `travel\-sample` WHERE type=$type LIMIT 1", "$type":"airline", "n1qlback": {"prepare": true}}
.fi
.IP "" 0
.P
Run the test load
.IP "" 4
.nf
cbc\-n1qlback \-\-num\-threads 5 \-\-queryfile queries\.txt
.fi
.IP "" 0
.SH "BUGS"
This command's options are subject to change\.
.SH "SEE ALSO"
cbc(1), cbc\-pillowfight(1), cbcrc(4)
.SH "HISTORY"
The \fBcbc\-n1qlback\fR tool was first introduced in libcouchbase 2\.4\.10


RELEASE_NOTES.markdown
# Release Notes

# 3.3.13 (2024-09-13)

* CCBC-1647: Handle server-side query timeout (status 1080).

* CCBC-1643: fix memory leak of pending buffers in IO context.

* CCBC-1645: do not use explicit initialization for OpenSSL 1.1+.

* Fix build with `-DLCB_DUMP_PACKETS=ON`

* Fix pkg-config for MacOS

# 3.3.12 (2024-03-02)

* CCBC-1636: Deallocate old packet when updating collection ID.
  `mcreq_renew_packet()` requires the caller to deallocate original copy, otherwise the memory will
  be only released by pipeline destructor.

* CCBC-1634: Fix reporting unresponsive nodes in `lcb_ping()`.
  * do not retry NOOP commands, as they might be routed to different pipeline, instead fail fast
    NOOPs to reflect network issues more precisely.
  * use pipeline address as ping entry identifier instead of socket address, as socket might not be
    existing (not connected) due to network failures.
  * `lcb_ping` still have report even when overall status is not `LCB_SUCCESS`, so cbc-ping should
    still try to print report instead just printing overall status code.

* CCBC-1630: Check collection id before storing packet to pipeline.
  Every time `check_collection_id()` is invoked, the caller should ensure that this function
  potentially is rewriting the packet, if it decides to insert/update encoded collection ID.

* CCBC-1627: Fix `bodylen` value when `ffextlen` (flexible frame extra length) is not zero.

# 3.3.11 (2023-12-21)

* CCBC-1618: update query error codes for dynamic authenticator. The dynamic authenticator is part of the internal API,
  and is not being used by default.

# 3.3.10 (2023-10-10)

* CCBC-1616: apply `wait_for_config` check for all pipelines. Previously, the `lcb_wait` function would wait for the
  pending configuration updates, but don't do it if the configuration update operation is being retried for some reason.
  Now, the operation also will not wait for pending configuration updates, and rather return from `lcb_wait` as soon as
  the operation completes. The old behaviour still works when `wait_for_config=true` is passed in connection string (or
  `LCB_CNTL_WAIT_FOR_CONFIG` set to non-zero value), in this case the library will wait for the configuration. This
  setting does not affect the mode, when the event loop is executed by the application, and without `lcb_wait`.

# 3.3.9 (2023-09-20)

* CCBC-1608: reduce timeout for idle HTTP connections to 1 second

* CCBC-1615: handle rate limit codes during bootstrap

* CCBC-1612: Improve recovery time during rebalance for upcoming Server 7.6.
    * do not throttle CCCP provider in faster failover mode.
    * try to refresh configuration in case of network errors to speed up the recovery process during failover.

* CCBC-1611: Handle `0x0d` (`ECONFIG_ONLY`) status code. Treat this status code as a signal to refresh configuration.
  The new or failed over nodes are set into config-only mode, where all data operatios will be failed with code 0x0d. It
  is possible that the SDK might be using stale configuration and send requests to the node, that is not part of the
  cluster anymore, so to work around this, the library will update the configuration and retry the operation.

* CCBC-1610: Fix memory issues when setting collection id in the cluster with mixed server versions, where some of the
  nodes do not support collections.

# 3.3.8 (2023-08-16)

* CCBC-1584: Update documentation on how to use collections with pillowfight

* CCBC-1607: Fix collection id encoding in mixed cluster

* CCBC-1602: Implement Faster Failover.

  This implements the set of protocol optimizations that help the SDK to save
  network traffic when tracking cluster topology. The feature will be only
  activated if the server supports it (7.6+).

CCBC-1603: Do not log if logger is not accessible in `iotssl_log_errors`.

CCBC-1599: Account NUL-byte when format IPv6 address (fixes potential invalid memory access).

## 3.3.7 (2023-05-11)

* CCBC-1596: replace unsafe sprintf with snprintf.

* CCBC-1597: Update threading example, reduce global state.

  - Json::Reader uses static global variable, this patch replaces calls
    to it with Json::CharReaderBuilder, which is re-entrant.

  - Constants for tracing system defined as mutable static strings, this
    patch replaces it with const static strings.

  - Updated examples for thread-safe usage is updated to SDK3 API and
    added them to the build pipeline


## 3.3.6 (2023-04-26)

* CCBC-1590: Always pick random node for HTTP services.

  It helps with certain edge cases, when the application might spawn a lot
  of processes, perform queries, so that first queries will be directed to
  the same node due to the absense of `srand()` call in the library.

* CCBC-1596: Fix various compiler warnings.

* CCBC-1592: Allow to generate more randomized bodies in pillowfight

  By default cbc-pillowfight pre-generates only one document body per
  selected size. New option `--random-body-pool-size` allows to control how
  many documents will be generated (default is 100).

  This fixes behaviour in the corner case when `--min-size` equals
  `--max-size` and allow still have many random bodies in this case.

* CCBC-1595: Fix building of the subdocument operation when `--subdoc`
  switch for pillowfight was used.

* pillowfight: use separate exptime switch for GET

  Do not share the same value of expiry for get operations.  Also it does
  not turn `GET` into `GET_WITH_TOUCH` if the --get-expiry is not being
  used.

## 3.3.5 (2023-03-09)

* CCBC-1545: handle `LCB_ERR_REQUEST_CANCELED` in ping callback

  If the instance is being destroyed, while the operations in flight, all
  these operations will be cancelled with error code
  `LCB_ERR_REQUEST_CANCELED`. Ping implementation should handle this error
  code and don't assume any of the objects (except response) be in valid
  state.

* CCBC-1586: force SASL PLAIN for TLS connections

* CCBC-1589: apply authenticator when passed to `lcb_create`


tools/cbc-n1qlback.cc
/* -*- Mode: C; tab-width: 4; c-basic-offset: 4; indent-tabs-mode: nil -*- */
/*
 *     Copyright 2015-2020 Couchbase, Inc.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License");
 *   you may not use this file except in compliance with the License.
 *   You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   Unless required by applicable law or agreed to in writing, software
 *   distributed under the License is distributed on an "AS IS" BASIS,
 *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *   See the License for the specific language governing permissions and
 *   limitations under the License.
 */

#include "config.h"
#include <sys/types.h>
#include <libcouchbase/couchbase.h>
#include <libcouchbase/vbucket.h>
#include <iostream>
#include <list>
#include <cassert>
#include <cstdio>
#include <cstdlib>
#include <cerrno>
#include <fstream>
#include <random>
#include <algorithm>
#include <stdexcept>
#ifndef WIN32
#include <pthread.h>
#include <unistd.h> // isatty()
#endif
#include "common/options.h"
#include "common/histogram.h"
#include "contrib/lcb-jsoncpp/lcb-jsoncpp.h"

using namespace cbc;
using namespace cliopts;
using std::cerr;
using std::endl;
using std::string;
using std::vector;

#ifndef _WIN32
static bool use_ansi_codes = true;
#else
static bool use_ansi_codes = false;
#endif

static void do_or_die(lcb_STATUS rc)
{
    if (rc != LCB_SUCCESS) {
        throw std::runtime_error(lcb_strerror_long(rc));
    }
}

struct Query {
    std::string payload{};
    bool prepare{false};
};

class Metrics
{
  public:
    Metrics() : n_rows(0), n_queries(0), n_errors(0), last_update(time(nullptr)), hg(nullptr)
    {
#ifndef _WIN32
        if (pthread_mutex_init(&m_lock, nullptr) != 0) {
            abort();
        }
#endif
        start_time = last_update;
    }

    size_t nerrors() const
    {
        return n_errors;
    }

    void update_row(size_t n = 1)
    {
        n_rows += n;
        update_display();
    }
    void update_done(size_t n = 1)
    {
        n_queries += n;
        update_display();
    }
    void update_error(size_t n = 1)
    {
        n_errors += n;
        update_display();
    }

    void update_timings(lcb_U64 duration)
    {
        if (hg != nullptr) {
            hg->record(duration);
        }
    }

#ifndef _WIN32
    static bool is_tty()
    {
        return isatty(STDOUT_FILENO);
    }
    void lock()
    {
        pthread_mutex_lock(&m_lock);
    }
    void unlock()
    {
        pthread_mutex_unlock(&m_lock);
    }
#else
    void lock() {}
    void unlock() {}
    static bool is_tty()
    {
        return false;
    }
#endif
    static void prepare_screen()
    {
        if (is_tty() && use_ansi_codes) {
            printf("\n\n\n");
        }
    }

    void prepare_timings()
    {
        if (hg == nullptr) {
            hg = new Histogram();
            hg->installStandalone(stdout);
        }
    }

  private:
    void update_display()
    {
        time_t now = time(nullptr);
        time_t duration = now - last_update;

        if (!duration) {
            return;
        }

        last_update = now;

        const char *prefix;
        const char *final_suffix;

        // Only use "ticker" style updates if we're a TTY and we have no
        // following timings.
        if (use_ansi_codes && is_tty() && hg == nullptr) {
            // Move up 3 cursors
            printf("\x1B[2A");
            prefix = "\x1B[K";
            final_suffix = "\r";
        } else {
            // Determine the total number of time
            unsigned total_duration = now - start_time;
            printf("\n"); // Clear line..
            printf("+%us\n", total_duration);
            prefix = "";
            final_suffix = "\n";
        }

        printf("%sQUERIES/SEC: %lu\n", prefix, (unsigned long)(n_queries / duration));
        printf("%sROWS/SEC:    %lu\n", prefix, (unsigned long)(n_rows / duration));
        printf("%sERRORS:      %lu%s", prefix, (unsigned long)n_errors, final_suffix);

        if (hg != nullptr) {
            hg->write();
        }
        fflush(stdout);

        n_queries = 0;
        n_rows = 0;
    }

    size_t n_rows;
    size_t n_queries;
    size_t n_errors;
    time_t last_update;
    time_t start_time;
    cbc::Histogram *hg;
#ifndef _WIN32
    pthread_mutex_t m_lock{};
#endif
};

class Configuration
{
  public:
    Configuration() : o_file("queryfile"), o_threads("num-threads"), o_errlog("error-log"), m_errlog(nullptr)
    {
        o_file.mandatory(true);
        o_file.description("Path to a file containing all the queries to execute. "
                           "Each line should contain the full query body");
        o_file.abbrev('f');

        o_threads.description("Number of threads to run");
        o_threads.abbrev('t');
        o_threads.setDefault(1);

        o_errlog.description("Path to a file containing failed queries");
        o_errlog.abbrev('e');
        o_errlog.setDefault("");
    }

    ~Configuration()
    {
        if (m_errlog != nullptr) {
            delete m_errlog;
            m_errlog = nullptr;
        }
    }
    void addToParser(Parser &parser)
    {
        parser.addOption(o_file);
        parser.addOption(o_threads);
        parser.addOption(o_errlog);
        m_params.addToParser(parser);
    }

    void processOptions()
    {
        std::ifstream ifs(o_file.const_result().c_str());
        if (!ifs.is_open()) {
            int ec_save = errno;
            string errstr(o_file.const_result());
            errstr += ": ";
            errstr += strerror(ec_save);
            throw std::runtime_error(errstr);
        }

        string curline;
        while (ifs.good()) {
            std::getline(ifs, curline);
            if (!curline.empty()) {
                Json::Value json;
                if (!parse_json(curline, json)) {
                    std::cerr << "Failed to parse query \"" << curline << "\" as JSON, skipping" << std::endl;
                    continue;
                }
                Query query{};
                Json::Value options = json.get("n1qlback", Json::nullValue);
                if (options.isObject()) {
                    Json::Value should_prepare = options.get("prepare", Json::nullValue);
                    if (should_prepare.isBool()) {
                        query.prepare = should_prepare.asBool();
                    }
                }
                json.removeMember("n1qlback");
                query.payload = Json::FastWriter().write(json);
                m_queries.emplace_back(query);
            }
        }
        std::cerr << "Loaded " << m_queries.size() << " queries "
                  << "from \"" << o_file.const_result() << "\"" << std::endl;
        if (m_params.useTimings()) {
            m_metrics.prepare_timings();
        }

        if (o_errlog.passed()) {
            m_errlog = new std::ofstream(o_errlog.const_result().c_str());
            if (!m_errlog->is_open()) {
                int ec_save = errno;
                string errstr(o_file.const_result());
                errstr += ": ";
                errstr += strerror(ec_save);
                throw std::runtime_error(errstr);
            }
            std::cerr << "Errors will be logged in \"" << o_errlog.const_result() << "\"" << std::endl;
        }
    }

    void set_cropts(lcb_CREATEOPTS *&opts)
    {
        m_params.fillCropts(opts);
    }
    const vector<Query> &queries() const
    {
        return m_queries;
    }
    size_t nthreads()
    {
        return o_threads.result();
    }
    std::ofstream *errlog()
    {
        return m_errlog;
    }

    Metrics &metrics()
    {
        return m_metrics;
    }

  private:
    vector<Query> m_queries;
    StringOption o_file;
    UIntOption o_threads;
    ConnParams m_params;
    StringOption o_errlog;
    std::ofstream *m_errlog;
    Metrics m_metrics{};
};

extern "C" {
static void n1qlcb(lcb_INSTANCE *, int, const lcb_RESPQUERY *resp);
}
extern "C" {
static void *pthrfunc(void *);
}

class ThreadContext;
struct QueryContext {
    lcb_U64 begin;
    bool received;      // whether any row was received
    ThreadContext *ctx; // Parent

    explicit QueryContext(ThreadContext *tctx) : begin(lcb_nstime()), received(false), ctx(tctx) {}
};

class ThreadContext
{
  public:
    void run()
    {
        while (!m_cancelled) {
            for (const auto &query : m_queries) {
                run_one_query(query);
            }
        }
    }

#ifndef _WIN32
    void start()
    {
        assert(m_thr == nullptr);
        m_thr = new pthread_t;
        int rc = pthread_create(m_thr, nullptr, pthrfunc, this);
        if (rc != 0) {
            throw std::runtime_error(strerror(rc));
        }
    }

    void join()
    {
        assert(m_thr != nullptr);
        void *arg = nullptr;
        pthread_join(*m_thr, &arg);
    }

    ~ThreadContext()
    {
        if (m_thr != nullptr) {
            join();
            delete m_thr;
            m_thr = nullptr;
        }
        lcb_cmdquery_destroy(m_cmd);
    }
#else
    void start()
    {
        run();
    }
    void join() {}
#endif

    void handle_response(const lcb_RESPQUERY *resp, QueryContext *ctx)
    {
        if (!ctx->received) {
            lcb_U64 duration = lcb_nstime() - ctx->begin;
            m_metrics.lock();
            m_metrics.update_timings(duration);
            m_metrics.unlock();

            ctx->received = true;
        }

        if (lcb_respquery_is_final(resp)) {
            lcb_STATUS rc = lcb_respquery_status(resp);
            if (rc != LCB_SUCCESS) {
                if (m_errlog != nullptr) {
                    const char *p;
                    size_t n;

                    lcb_cmdquery_encoded_payload(m_cmd, &p, &n);
                    std::stringstream ss;
                    ss.write(p, n);
                    ss << endl;
                    lcb_respquery_row(resp, &p, &n);
                    ss.write(p, n);
                    log_error(rc, ss.str().c_str(), ss.str().size());
                } else {
                    log_error(rc, nullptr, 0);
                }
            }
        } else {
            last_nrow++;
        }
    }

    ThreadContext(Metrics &metrics, lcb_INSTANCE *instance, const vector<Query> &initial_queries, std::ofstream *errlog)
        : m_instance(instance), last_nrow(0), m_cmd(nullptr), m_metrics(metrics), m_cancelled(false), m_thr(nullptr),
          m_errlog(errlog)
    {
        lcb_cmdquery_create(&m_cmd);
        lcb_cmdquery_callback(m_cmd, n1qlcb);

        // Shuffle the list
        m_queries = initial_queries;
        std::random_device rd;
        std::mt19937 g(rd());
        std::shuffle(m_queries.begin(), m_queries.end(), g);
    }

  private:
    void log_error(lcb_STATUS err, const char *info, size_t ninfo)
    {
        size_t erridx;
        m_metrics.lock();
        m_metrics.update_error();
        erridx = m_metrics.nerrors();
        m_metrics.unlock();

        if (m_errlog != nullptr) {
            std::stringstream ss;
            ss << "[" << erridx << "] " << lcb_strerror_short(err) << endl;
            if (ninfo) {
                ss.write(info, ninfo);
                ss << endl;
            }
            *m_errlog << ss.str();
            m_errlog->flush();
        }
    }

    void run_one_query(const Query &query)
    {
        // Reset counters
        last_nrow = 0;

        lcb_cmdquery_payload(m_cmd, query.payload.c_str(), query.payload.size());
        lcb_cmdquery_adhoc(m_cmd, !query.prepare);

        // Set up our context
        QueryContext qctx(this);

        lcb_STATUS rc = lcb_query(m_instance, &qctx, m_cmd);
        if (rc != LCB_SUCCESS) {
            log_error(rc, query.payload.c_str(), query.payload.size());
            lcb_tick_nowait(m_instance);
        } else {
            lcb_wait(m_instance, LCB_WAIT_DEFAULT);
            m_metrics.lock();
            m_metrics.update_row(last_nrow);
            m_metrics.update_done(1);
            m_metrics.unlock();
        }
    }

    lcb_INSTANCE *m_instance;
    vector<Query> m_queries;
    size_t last_nrow;
    lcb_CMDQUERY *m_cmd;
    Metrics &m_metrics;
    volatile bool m_cancelled;
#ifndef _WIN32
    pthread_t *m_thr;
#else
    void *m_thr;
#endif
    std::ofstream *m_errlog;
};

static void n1qlcb(lcb_INSTANCE *, int, const lcb_RESPQUERY *resp)
{
    QueryContext *qctx;
    lcb_respquery_cookie(resp, (void **)&qctx);
    qctx->ctx->handle_response(resp, qctx);
}

static void *pthrfunc(void *arg)
{
    reinterpret_cast<ThreadContext *>(arg)->run();
    return nullptr;
}

static bool instance_has_n1ql(lcb_INSTANCE *instance)
{
    // Check that the instance supports N1QL
    lcbvb_CONFIG *vbc;
    do_or_die(lcb_cntl(instance, LCB_CNTL_GET, LCB_CNTL_VBCONFIG, &vbc));

    int sslmode = 0;
    lcbvb_SVCMODE svcmode = LCBVB_SVCMODE_PLAIN;

    do_or_die(lcb_cntl(instance, LCB_CNTL_GET, LCB_CNTL_SSL_MODE, &sslmode));

    if (sslmode & LCB_SSL_ENABLED) {
        svcmode = LCBVB_SVCMODE_SSL;
    }

    int hix = lcbvb_get_randhost(vbc, LCBVB_SVCTYPE_QUERY, svcmode);
    return hix > -1;
}

static void real_main(int argc, char **argv)
{
    Configuration config;

    Parser parser;
    config.addToParser(parser);
    parser.parse(argc, argv);

    vector<ThreadContext *> threads;
    vector<lcb_INSTANCE *> instances;

    lcb_CREATEOPTS *cropts = nullptr;
    config.set_cropts(cropts);
    config.processOptions();

    for (size_t ii = 0; ii < config.nthreads(); ii++) {
        lcb_INSTANCE *instance;
        do_or_die(lcb_create(&instance, cropts));
        do_or_die(lcb_connect(instance));
        lcb_wait(instance, LCB_WAIT_DEFAULT);
        do_or_die(lcb_get_bootstrap_status(instance));

        if (ii == 0 && !instance_has_n1ql(instance)) {
            throw std::runtime_error("Cluster does not support N1QL!");
        }

        auto *cx = new ThreadContext(config.metrics(), instance, config.queries(), config.errlog());
        threads.push_back(cx);
        instances.push_back(instance);
    }
    lcb_createopts_destroy(cropts);

    Metrics::prepare_screen();

    for (auto &thread : threads) {
        thread->start();
    }
    for (auto &thread : threads) {
        thread->join();
    }
    for (auto &instance : instances) {
        lcb_destroy(instance);
    }
}

int main(int argc, char **argv)
{
    try {
        real_main(argc, argv);
        return 0;
    } catch (std::exception &exc) {
        cerr << exc.what() << endl;
        exit(EXIT_FAILURE);
    }
}
